### ISSUE TITLE
Saving NSDate with FMDB, but they are not equal when load from FMDB.

### ISSUE THREAD

**Issue by heistings** [ccgus/fmdb#641]
I saved a NSDate as BLOB type, and load with `dateForColumn` method.

Something weird happened, the two date object are not equal, though the description are the same:

```
XCTAssert([originDate isEqualToDate:date]); // failed
XCTAssert([date.description isEqualToString:originDate.description]); // success
```

**Comment by ccgus** [comment]
Can you modify the FMDBReportABugFunction to show your example failing? ( see https://github.com/ccgus/fmdb#reporting-bugs )

**Comment by robertmryan** [comment]
I'm not sure what you mean by "saved ... as BLOB". How did you get `NSData` representation of your `NSDate`?

I wonder if you meant that you inserted `NSDate` into a column that you had defined as a `BLOB`. I know that sounds like the same thing, but it isn't. SQLite doesn't enforce column types and FMDB could have saved your date as a `double`, equivalent to`timeIntervalSince1970`. Show us how you "saved ... as BLOB". Or you can examine the database in an external tool (e.g. command line `sqlite3` program; or whatever SQLite tool you like) and confirm precisely what was stored in that column in the table.

---

Just so you know, if it was the case that it used `timeIntervalSince1970`, there is a known issue with that pattern. If you grab the `timeIntervalSince1970`, build a `NSDate` from that `double`, the resulting date will not equal the original one:

```objc
NSDate *date = [NSDate date];
NSTimeInterval interval = [date timeIntervalSince1970];
NSDate *date2 = [NSDate dateWithTimeIntervalSince1970:interval];

if (![date isEqual:date2]) {
   NSLog(@"not equal!!!");
}
```

I believe this is a byproduct of floating point errors introduced when converting `NSDate` to and from the `double`.

For the record, you can use `NSKeyedArchiver` to really store `NSDate` as a `BLOB` and not suffer from this problem. But it's horribly inefficient pattern.

---

Bottom line, show us how you "saved ... as BLOB", and we can advise you further.

**Comment by bamx23** [comment]
This is because of NSDate storing time delta [since 2001](https://developer.apple.com/documentation/foundation/nsdate/1417376-timeintervalsincereferencedate), not 1970. This fact leads to floating point errors. If you store`[date timeIntervalSinceReferenceDate]`, you get an exact date back.
